R/kelp.R

```diff
@@ -0,0 +1,110 @@
+#' Create a kelp object for interacting with SeaweedFS.
+#'
+#' This is a slightly higher level abstraction than [seaweed_master()] and
+#' [seaweed_volume()].
+#'
+#' @keywords export
+#' @importFrom R6 R6Class
+kelp <- R6::R6Class(
+  "kelp",
+  cloneable = FALSE,
+
+  public = list(
+    #' @field master A [seaweed_master] object for sending requests
+    master = NULL,
+
+    #' @description
+    #' Create client object for sending http requests to seaweed
+    #'
+    #' @param seaweed_url Root URL of Seaweed
+    #'
+    #' @return A new `kelp` object
+    initialize = function(seaweed_url) {
+      self$master <- seaweed_master$new(seaweed_url)
+    },
+
+    #' @description
+    #' Upload file to SeaweedFS
+    #'
+    #' @param path Path to file to be uploaded
+    #' @param collection Collection name, acts as a namespace for files.
+    #'
+    #' @return The uploaded file ID.
+    upload_file = function(path, collection = NULL) {
+      key <- self$master$assign(collection = collection)
+      volume <- seaweed_volume$new(key$publicUrl)
+      volume$upload_file(key$fid, path)
+      key$fid
+    },
+
+    #' @description
+    #' Download file from SeaweedFS
+    #'
+    #' @param fid SeaweedFS file ID to read
+    #' @param path Path to download file to
+    #' @param collection Optional collection name this file belongs to.
+    #'
+    #' @return The file contents
+    download_file = function(fid, path = tempfile(), collection = NULL) {
+      volumes <- self$master$lookup(fid, collection)
```

Comment:
What happens on failure to lookup the fid here?

Reply:
This will either error with message `Unknown volume id <volume_id>` or `Client error: (404) Not Found` if the volume part of the id is not found or if subsequently the file can't be downloaded. It's just directly passing the message through from SeaweedFS.